File: js/pythonCode/isWin.py

NONE  = 0
RED   = 1
GREEN = 2
BLUE  = 3

# Assume input int[4][4]
def isWin(board):
    for r in range(0, 4):
        for c in range(0, 4):
          # I changed the next three lines
          tempVal = threeInRow(board, r, c)
          if tempVal != 0:
            return tempVal
    return NONE

def threeInRow(board, row, col):
    currColor = board[row][col]
    if currColor != NONE and (inDiagonal(board, row, col, 3, currColor) or inRow(board, row, col, 3, currColor) or inCol(board, row, col, 3, currColor)):
        return currColor # I changed this
    else:
        return NONE # I changed This

def inDiagonal(board, row, col, distance, color):
    if distance == 0:
        return True 
    
    if row < 0 or row > 3 or col < 0 or col > 3:
        return False

    if board[row][col] == color and inDiagonal(board, row + 1, col + 1, distance - 1, color):
        return True 
    else:
        return False

def inRow(board, row, col, distance, color):
    if distance == 0:
        return True 
    
    if row < 0 or row > 3 or col < 0 or col > 3:
        return False
    
    if board[row][col] == color and inRow(board, row + 1, col, distance - 1, color):
        return True 
    else:
        return False

def inCol(board, row, col, distance, color):
    if distance == 0:
        return True 
    
    if row < 0 or row > 3 or col < 0 or col > 3:
        return False
    
    if board[row][col] == color and inRow(board, row, col + 1, distance - 1, color):
        return True 
    else:
        return False
'''
Code Fails Diagonal Test

test3 = [[1,0,0,2],[0,0,2,0],[1,2,0,0],[2,1,0,0]]
print(isWin(test3))

Expected Output: 2
Actual Output: 0 (in python), null (in JS)



'''

File: js/pythonCode/isWin.js

const NONE  = 0
const RED   = 1
const GREEN = 2
const BLUE  = 3

// Assume input int[4][4]
function isWin(board){
    for (let r = 0; r < 4; r++){
        for (let c = 0; c < 4; c++){
            
            let tempVal = threeInRow(board, r, c);

            if (tempVal != 0){
                return tempVal;
            }
        }
    }
    return null;
}

function threeInRow(board, row, col){
    currColor = board[row][col]
    if (currColor != null && (inDiagonal(board, row, col, 3, currColor) || inRow(board, row, col, 3, currColor) || inCol(board, row, col, 3, currColor))){
        return currColor;
    } 
    else{
        return null; // I changed This
    }
        
}

function inDiagonal(board, row, col, distance, color){
    if (distance == 0){
        return true;
    }
         
    
    if (row < 0 || row > 3 || col < 0 || col > 3){
        return false;
    }
        

    if (board[row][col] == color && inDiagonal(board, row + 1, col + 1, distance - 1, color)){
        return true;
    }
   
    else{
        return false;
    }
        
}
    

function inRow(board, row, col, distance, color){
    if (distance == 0){
        return true;
    } 
    
    if (row < 0 || row > 3 || col < 0 || col > 3){
        return false;
    }
    
    if (board[row][col] == color && inRow(board, row + 1, col, distance - 1, color)){
        return true;
    }
         
    else{
        return false;
    }
}
function inCol(board, row, col, distance, color){
    if (distance == 0){
        return true;
    }  
    
    if (row < 0 || row > 3 || col < 0 || col > 3){
        return false;
    }
        
    
    if (board[row][col] == color && inRow(board, row, col + 1, distance - 1, color)){
        return true;
    }
         
    else{
        return false;
    }
}